Firestore/core/src/firebase/firestore/remote/stream_operation.cc

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/stream_operation.h"
+
+#include "Firestore/core/src/firebase/firestore/remote/grpc_stream.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+using util::AsyncQueue;
+
+StreamOperation::StreamOperation(GrpcStream* stream,
+                                 grpc::GenericClientAsyncReaderWriter* call,
+                                 AsyncQueue* firestore_queue)
+    : observer_{stream}, call_{call}, firestore_queue_{firestore_queue} {
+}
+
+void StreamOperation::UnsetObserver() {
+  firestore_queue_->VerifyIsCurrentQueue();
+
+  observer_ = nullptr;
+}
+
+void StreamOperation::Execute() {
+  firestore_queue_->VerifyIsCurrentQueue();
+
+  DoExecute(call_);
+}
+
+void StreamOperation::WaitUntilOffQueue() {
+  firestore_queue_->VerifyIsCurrentQueue();
+
+  if (!off_queue_future_.valid()) {
+    off_queue_future_ = off_queue_.get_future();
+  }
+  return off_queue_future_.wait();
+}
+
+std::future_status StreamOperation::WaitUntilOffQueue(
+    std::chrono::milliseconds timeout) {
+  firestore_queue_->VerifyIsCurrentQueue();
+
+  if (!off_queue_future_.valid()) {
+    off_queue_future_ = off_queue_.get_future();
+  }
+  return off_queue_future_.wait_for(timeout);
+}
+
+void StreamOperation::Complete(bool ok) {
+  // This mechanism allows `GrpcStream` to know when the operation is off the
+  // GRPC completion queue (and thus this operation no longer requires the
+  // underlying GRPC objects to be valid).
+  off_queue_.set_value();
+
+  firestore_queue_->Enqueue([this, ok] {
```

Comment:
I'm torn about this and I'm not sure this is good advice, but think about this:

It seems that `StreamOperations` are already threading aware because of the promise they complete. `StreamOperation` wouldn't need to know about the async queue at all if the GrpcStream itself handled this detail. The trouble is that we currently null out `observer_` on the worker thread, maybe we could just avoid that?

Consider if `GrpcStream` had a method like this:

```c++
void GrpcStream::Complete(StreamOperation* operation, bool ok) {
  firestore_queue->Enqueue([this, operation, ok] {
    if (observer_) {
      RemoveOperation(operation);

      if (ok) {
        operation->CompleteSuccessfully(this);  // better name ideas welcome.
      } else {
        OnOperationFailed();
      }
    }
    delete operation;
  });
}
```
This would have a few benefits:
  * The GrpcStream would be calling both `new` and `delete`
  * The `StreamOperation`s could be async queue agnostic

Reply:
I might misunderstand the idea, but it would seem to be used like this:
```cpp
// Called by a dedicated executor polling gRPC queue
void StreamOperation::MarkReady(bool ok) {
  stream_->MarkReady(this, ok); // Line A
  // ...presumably complete the promise, etc...
}

// ...

// Called on Firestore queue
void GrpcStream::FastFinishOperationsBlocking() {
  // ...
  for (auto o : operations_) {
    o->UnsetStream(); // Line B (presume void StreamOperation::UnsetStream() { stream_ = nullptr; })
  }
  // ...
}
```
I might be wrong, but it seems line A and line B would have a race on `stream_` -- line A uses the pointer while line B nulls it.